enhancements/node-tuning/pao-in-nto.md

```diff
@@ -0,0 +1,280 @@
+
+- [performance-addon-must-gather](https://catalog.redhat.com/software/containers/openshift4/performance-addon-operator-must-gather-rhel8/5ee0bfa4d70cc56cea83827f?container-tabs=overview&gti-tabs=registry-tokens)
+
+  The must gather image [enhances](https://github.com/openshift-kni/performance-addon-operators/tree/master/must-gather) the main OCP one and [collects](https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-about-gathering-data_cnf-master) some latency tuning configuration and data that are needed for debugging.
+
+- functional tests are also vendored into
+[cnf-tests](https://catalog.redhat.com/software/containers/openshift4/cnf-tests-rhel8/5e9d81f65a134668769d560b?container-tabs=overview&gti-tabs=registry-tokens)
+that can be used to
+[validate customer deployments](https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-performing-end-to-end-tests-for-platform-verification_cnf-master) (this includes other aspects like SCTP, PTP, SR-IOV as well).
+
+## Motivation
+
+The Telco team maintains its own build and release pipelines and releases via a [Telco specific errata](https://red.ht/3boqjeS) that however targets the OpenShift product. So from a customer perspective PAO is part of OCP.
+
+- More and more customers are using PAO to configure their clusters for low latency or are being told to do so. The Telco team has so far managed to release PAO on the same day as OCP to maintain the illusion it is part of the product, but it costs a LOT of time to maintain the downstream build process.
+
+- The OCP [gcp-rt](https://prow.ci.openshift.org/?type=periodic&job=*gcp-rt*) lane tests OCP deployment on top of GCP with the real-time kernel enabled, but it only works [thanks to some pieces of PAO tuning](https://github.com/openshift/release/pull/18870). However the tuning is a pre-rendered snippet, not linked to the operator, because the deployment does not know about PAO.
+
+- PAO is closely tied to some OCP components
+([Cluster Node Tuning Operator](https://github.com/openshift/cluster-node-tuning-operator))
+and depends on the OCP version. So we release PAO for every OCP minor version (since 4.4) and together with some of the OCP Z-streams.
+The upgrade procedure is tricky though, first OCP needs to be upgraded and then PAO via an OLM channel change.
+This is needed to make sure the necessary functionality is present in the running NTO.
+
+- The new [Workload partitioning](https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md)
+feature seeks to use PAO as the configuration tool.
+However Workload partitioning needs to be enabled at install phase and PAO CRDs are not yet available at that time.
+[A "hacky" procedure](https://github.com/openshift/enhancements/pull/753) is needed to overcome this.
```

Comment:
There's a strong desire to reduce the number of components delivered as part of the OCP release payload, and have more of them installed via OLM. There are several benefits to that approach, including decoupling release schedules, shrinking the payload, and making it possible to reduce the API surface area for alternative topologies. The PAO seems like a strong candidate for OLM delivery, especially since that's how it already works.

If we do add it to the payload, NTO seems like a reasonable "home" (although I could also see moving NTO to OLM delivery). But I'm not yet convinced it's necessary to make this change, based on the motivation described here.

Regarding integration with the installer, it would be useful to have additional requirements added to https://issues.redhat.com/browse/OLM-2232, if the plans there don't already cover the use case for PAO.

Reply:
There is a strong coupling to NTO features and sometimes to a specific kernel bug fixes delivered in RHCOS. 

One of the biggest gaps in OLM for such infrastructure operators (apart from the 1st day install described by OLM-2232) is no support for atomic upgrades and proper dependency mechanism that would support multiple Z-streams.

The merge with NTO was proposed to not increase the amount of installed and running processes indeed.

Moving NTO itself to OLM sounds tricky to me without a way to setup at least the openshift-node tuned profile in RHCOS, this is something I consider important for the platform as a whole (@jmencak do you have the original design for NTO?).

Reply:
@MarSik, the original design document/proposal is a [google doc](https://github.com/openshift/enhancements/pull/867#issuecomment-897607025). I believe the audience for NTO is [a lot wider](https://github.com/redhat-performance/tuned/tree/master/profiles) than for PAO, but PAO definitely helps where low-latency tuning is needed. The main aim behind NTO was to concentrate node tuning into one place and take advantage of years of experience transferred into [workload-specific profiles](https://github.com/redhat-performance/tuned/tree/master/profiles).  `TuneD` has been an integral part of `RHEL` for many years and customers have come to expect it there.